lib/validators/AlgorandTxValidator.js

```diff
@@ -1,47 +1,101 @@
-        if (hwTokensBalance < txTokensAmount) {
-          const errorMessage = `The Hot Wallet has insufficient tokens to transfer (${hwTokensBalance} < ${txTokensAmount})`
-          console.log(errorMessage)
-          return { valid: false, markAs: "cancelled", error: errorMessage }
-        }
+      if (isAmountValid && isAppIndexValid) {
+        return { valid: true }
+      } else {
+        return { valid: false, markAs: this.markAs, error: this.error }
       }
-      return { valid: true }
     } catch (err) {
       console.error(err)
       return { valid: false, markAs: "failed", error: `Unknown error: ${err}` }
     }
   }
+
+  // Is AppIndex for the current App?
+  isAppIndexValid(transaction, transactableType) {
+    if (!this.envs.optInApp || this.skipAppIndexTypes.includes(transactableType) ||
+      this.envs.optInApp && transaction.appIndex === this.envs.optInApp ) {
+        return true
+    } else {
+      this.markAs = "failed"
+      this.error = "The transaction is not for configured App."
+      return false
+    }
+  }
+
+  async isAmountValid({ transaction, transactableType, hotWalletAddress }) {
+    let isValid = true
+    const transferMethodId = "0x7472616e73666572"
+    if (this.envs.optInApp && transaction.appArgs?.[0] === transferMethodId) {
+      const txTokensAmount = parseInt(transaction.appArgs[1])
+      isValid = await this.isMaxAmountNotExceeded(txTokensAmount) &&
+        await this.isTokenBalanceEnough({ txTokensAmount, hotWalletAddress })
+    } else if (transactableType === 'FundOptIn') {
+      const txCoinsAmount = parseInt(transaction.amount)
+      isValid = await this.isCoinsBalanceEnough({ txCoinsAmount, hotWalletAddress })
+    }
+    return isValid
+  }
+
+  async isMaxAmountNotExceeded(txTokensAmount) {
+    if (!this.envs.maxAmountForTransfer || txTokensAmount <= this.envs.maxAmountForTransfer) {
+      return true
+    } else {
+      const errorMessage = `The transaction's transfer is too large (${txTokensAmount}). Max amount is ${this.envs.maxAmountForTransfer}`
+      console.log(errorMessage)
+      this.markAs = "failed"
+      this.error = errorMessage
+      return false
+    }
+  }
+
+  async isTokenBalanceEnough({ txTokensAmount, hotWalletAddress }) {
+    const hwTokensBalance = await this.blockchain.getTokenBalance(hotWalletAddress)
+
+    if (hwTokensBalance >= txTokensAmount) {
+      return true
+    } else {
+      const errorMessage = `The Hot Wallet has insufficient tokens to transfer (${hwTokensBalance} < ${txTokensAmount})`
+      console.log(errorMessage)
+      this.markAs = "cancelled"
+      this.error = errorMessage
+      return false
+    }
+  }
+
+  async isCoinsBalanceEnough({ txCoinsAmount, hotWalletAddress }) {
+    const hwCoinsBalance = await this.blockchain.getAlgoBalanceForHotWallet(hotWalletAddress)
+    const txAmount = new BigNumber(txCoinsAmount).dividedBy(1000000)
```

Comment:
I think it would be better representing the balance in base units here, rather than dividing amount to get value in ALGO.

This way we will always work with integers.